File: server/toolkit.ts
import { execFile } from 'node:child_process'
import { constants } from 'node:fs'
import { access, copyFile, mkdir, readdir, readlink, rm, symlink } from 'node:fs/promises'
import path from 'node:path'
import process from 'node:process'
import { promisify } from 'node:util'

export async function toolkitEnvironment(home: string, base: NodeJS.ProcessEnv = process.env): Promise<NodeJS.ProcessEnv> {
  if (!base.LEO_TOOLKIT_DIR)
    return { ...base }
  const directory = base.LEO_TOOLKIT_DIR
  const rustup = path.join(home, '.rustup')
  const cargo = path.join(home, '.cargo')
  await mkdir(path.join(rustup, 'toolchains'), { recursive: true })
  await mkdir(path.join(cargo, 'bin'), { recursive: true })
  for (const folder of ['update-hashes', 'downloads', 'tmp'])
    await mkdir(path.join(rustup, folder), { recursive: true })
  for (const hash of await readdir(path.join(directory, 'rustup/update-hashes'))) {
    await copyFile(path.join(directory, 'rustup/update-hashes', hash), path.join(rustup, 'update-hashes', hash), constants.COPYFILE_EXCL).catch((error: NodeJS.ErrnoException) => {
      if (error.code !== 'EEXIST')
        throw error
    })
  }
  const link = async (source: string, target: string) => {
    await symlink(source, target).catch((error: NodeJS.ErrnoException) => {
      if (error.code !== 'EEXIST')
        throw error
    })
  }
  // A replacement image may no longer contain an intermediate tool version.
  // Remove only dangling links to our image, so that version can be installed locally.
  const prune = async (folder: string) => {
    for (const name of await readdir(folder)) {
      const file = path.join(folder, name)
      const target = await readlink(file).catch(() => '')
      if ((target.startsWith('/usr/local/share/mise/installs/') || target.startsWith(path.join(directory, 'rustup/toolchains/')))
        && !await access(file).then(() => true).catch(() => false)) {
        await rm(file)
      }
    }
  }
  await prune(path.join(rustup, 'toolchains'))
  const installs = path.join(home, '.local/share/mise/installs')
  // These backends discover idiomatic version files from the install registry.
  // Register their shared versions locally so new project pins install in the home.
  for (const tool of ['node', 'pnpm', 'python', 'go']) {
    const source = path.join('/usr/local/share/mise/installs', tool)
    const target = path.join(installs, tool)
    await mkdir(target, { recursive: true })
    await prune(target)
    await copyFile(path.join(source, '.mise.backend.toml'), path.join(target, '.mise.backend.toml'), constants.COPYFILE_EXCL).catch((error: NodeJS.ErrnoException) => {
      if (error.code !== 'EEXIST' && error.code !== 'ENOENT')
        throw error
    })
    for (const version of await readdir(source)) {
      if (!/^\d+\.\d+\.\d+$/.test(version))
        continue
      await link(path.join(source, version), path.join(target, version))
    }
  }
  const shims = path.join(home, '.local/share/mise/shims')
  await mkdir(shims, { recursive: true })
  for (const name of await readdir('/usr/local/share/mise/shims'))
    await link('/usr/local/bin/mise', path.join(shims, name))
  for (const name of await readdir(path.join(directory, 'rustup/toolchains')))
    await link(path.join(directory, 'rustup/toolchains', name), path.join(rustup, 'toolchains', name))
  for (const name of await readdir(path.join(directory, 'cargo/bin')))
    await link(path.join(directory, 'cargo/bin', name), path.join(cargo, 'bin', name))
  const settings = path.join(rustup, 'settings.toml')
  if (!await access(settings).then(() => true).catch(() => false)) {
    await copyFile(path.join(directory, 'rustup/settings.toml'), settings, constants.COPYFILE_EXCL).catch((error: NodeJS.ErrnoException) => {
      if (error.code !== 'EEXIST')
        throw error
    })
  }
  const env = {
    ...base,
    HOME: home,
    RUSTUP_HOME: rustup,
    CARGO_HOME: cargo,
    PATH: [path.join(home, '.local/share/mise/shims'), '/usr/local/share/mise/shims', path.join(cargo, 'bin'), base.PATH].filter(Boolean).join(path.delimiter),
  }
  await promisify(execFile)('/usr/local/bin/mise', ['reshim'], { cwd: '/tmp', env, timeout: 30000 })
  return env
}

File: backend/src/toolkit.rs
use crate::{
    error::{Error, Result},
    process::{Environment, bounded_output, command},
};
use std::{path::Path, time::Duration};
async fn copy_new(source: &Path, target: &Path, optional: bool) -> Result<()> {
    use std::os::unix::fs::PermissionsExt;
    let mut source = match tokio::fs::File::open(source).await {
        Ok(source) => source,
        Err(e) if optional && e.kind() == std::io::ErrorKind::NotFound => return Ok(()),
        Err(e) => return Err(e.into()),
    };
    let mode = source.metadata().await?.permissions().mode();
    let mut target = match tokio::fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .mode(mode)
        .open(target)
        .await
    {
        Ok(target) => target,
        Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => return Ok(()),
        Err(e) => return Err(e.into()),
    };
    tokio::io::copy(&mut source, &mut target).await?;
    Ok(())
}
async fn link(source: &Path, target: &Path) -> Result<()> {
    match tokio::fs::symlink(source, target).await {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => Ok(()),
        Err(e) => Err(e.into()),
    }
}
async fn names(path: &Path) -> Result<Vec<String>> {
    let mut entries = tokio::fs::read_dir(path).await?;
    let mut names = Vec::new();
    while let Some(entry) = entries.next_entry().await? {
        names.push(
            entry
                .file_name()
                .to_str()
                .ok_or_else(|| Error::internal("Invalid toolkit filename"))?
                .to_owned(),
        );
    }
    Ok(names)
}
async fn prune(path: &Path, directory: &Path) -> Result<()> {
    for name in names(path).await? {
        let file = path.join(name);
        if let Ok(target) = tokio::fs::read_link(&file).await
            && (target.starts_with("/usr/local/share/mise/installs")
                || target.starts_with(directory.join("rustup/toolchains")))
            && !tokio::fs::try_exists(&file).await?
        {
            tokio::fs::remove_file(file).await?;
        }
    }
    Ok(())
}
pub async fn environment(home: &Path, mut env: Environment) -> Result<Environment> {
    let Some(directory) = env.get("LEO_TOOLKIT_DIR").cloned() else {
        return Ok(env);
    };
    let directory = Path::new(&directory);
    let rustup = home.join(".rustup");
    let cargo = home.join(".cargo");
    for path in [
        rustup.join("toolchains"),
        cargo.join("bin"),
        rustup.join("update-hashes"),
        rustup.join("downloads"),
        rustup.join("tmp"),
    ] {
        tokio::fs::create_dir_all(path).await?;
    }
    for name in names(&directory.join("rustup/update-hashes")).await? {
        copy_new(
            &directory.join("rustup/update-hashes").join(&name),
            &rustup.join("update-hashes").join(name),
            false,
        )
        .await?;
    }
    prune(&rustup.join("toolchains"), directory).await?;
    let installs = home.join(".local/share/mise/installs");
    for tool in ["node", "pnpm", "python", "go", "rust"] {
        let source = Path::new("/usr/local/share/mise/installs").join(tool);
        let target = installs.join(tool);
        tokio::fs::create_dir_all(&target).await?;
        prune(&target, directory).await?;
        copy_new(
            &source.join(".mise.backend.toml"),
            &target.join(".mise.backend.toml"),
            true,
        )
        .await?;
        for version in names(&source).await? {
            let parts = version.split('.').collect::<Vec<_>>();
            if parts.len() == 3
                && parts
                    .iter()
                    .all(|p| !p.is_empty() && p.bytes().all(|b| b.is_ascii_digit()))
            {
                link(&source.join(&version), &target.join(version)).await?;
            }
        }
    }
    let shims = home.join(".local/share/mise/shims");
    tokio::fs::create_dir_all(&shims).await?;
    for name in names(Path::new("/usr/local/share/mise/shims")).await? {
        link(Path::new("/usr/local/bin/mise"), &shims.join(name)).await?;
    }
    for name in names(&directory.join("rustup/toolchains")).await? {
        link(
            &directory.join("rustup/toolchains").join(&name),
            &rustup.join("toolchains").join(name),
        )
        .await?;
    }
    for name in names(&directory.join("cargo/bin")).await? {
        let target = cargo.join("bin").join(&name);
        if tokio::fs::read_link(&target)
            .await
            .is_ok_and(|p| p.starts_with(directory.join("cargo/bin")))
        {
            tokio::fs::remove_file(&target).await?;
        }
        if name == "rustup" {
            copy_new(&directory.join("cargo/bin/rustup"), &target, false).await?;
        } else {
            link(&cargo.join("bin/rustup"), &target).await?;
        }
    }
    copy_new(
        &directory.join("rustup/settings.toml"),
        &rustup.join("settings.toml"),
        false,
    )
    .await?;
    env.insert("HOME".into(), home.to_string_lossy().into_owned());
    env.insert("RUSTUP_HOME".into(), rustup.to_string_lossy().into_owned());
    env.insert("CARGO_HOME".into(), cargo.to_string_lossy().into_owned());
    env.insert(
        "PATH".into(),
        format!(
            "{}:/usr/local/share/mise/shims:{}:{}",
            shims.display(),
            cargo.join("bin").display(),
            env.get("PATH")
                .map(String::as_str)
                .unwrap_or("/usr/local/bin:/usr/bin:/bin")
        ),
    );
    for args in [vec!["reshim".into()], vec!["env".into(), "--json".into()]] {
        let output = bounded_output(
            command("/usr/local/bin/mise", &args, &env, Some(Path::new("/tmp"))),
            Duration::from_secs(30),
            100000,
        )
        .await?;
        if !output.success {
            return Err(Error::new(503, "Unable to prepare the agent toolkit."));
        }
    }
    Ok(env)
}
